qadence/analog/addressing.py

```diff
@@ -0,0 +1,138 @@
+
+    # maximum amplitude can also be chosen as a variational parameter if needed
+    max_amp: float | torch.Tensor | Parameter = DEFAULT_MAX_AMPLITUDE
+
+    # maximum detuning can also be chosen as a variational parameter if needed
+    max_det: float | torch.Tensor | Parameter = DEFAULT_MAX_DETUNING
+
+    # weight constraint
+    weight_constraint: WeightConstraint = WeightConstraint.NORMALIZE
+
+    def _normalize_weights(self) -> None:
+        self.weights_amp = {
+            k: Parameter(v) if not isinstance(v, Parameter) else abs(v)
+            for k, v in self.weights_amp.items()
+        }
+        sum_weights_amp = sum(list(self.weights_amp.values()))
+        self.weights_amp = {k: v / sum_weights_amp for k, v in self.weights_amp.items()}
+
+        self.weights_det = {
+            k: Parameter(v) if not isinstance(v, Parameter) else abs(v)
+            for k, v in self.weights_det.items()
+        }
+        sum_weights_det = sum(list(self.weights_det.values()))
+        self.weights_det = {k: v / sum_weights_det for k, v in self.weights_det.items()}
+
+    def _restrict_weights(self) -> None:
+        self.weights_amp = {
+            k: v * (sigmoid(v, 20, 0.0) - sigmoid(v, 20.0, -1.0))
+            for k, v in self.weights_amp.items()
+        }
+        self.weights_det = {
+            k: v * (sigmoid(v, 20.0, 0.0) - sigmoid(v, 20.0, -1.0))
+            for k, v in self.weights_det.items()
+        }
+
+    def _restrict_max_vals(self) -> None:
+        self.max_amp = self.max_amp * (
+            sympy.Heaviside(self.max_amp) - sympy.Heaviside(self.max_amp - DEFAULT_MAX_AMPLITUDE)
+        )
+        self.max_det = self.max_det * (
+            sympy.Heaviside(self.max_det) - sympy.Heaviside(self.max_det - DEFAULT_MAX_DETUNING)
+        )
+
+    def __post_init__(self) -> None:
+        # validate weights
+        if all([not isinstance(v, Parameter) for v in self.weights_amp.values()]):
+            if not torch.isclose(
+                torch.tensor(list(self.weights_amp.values())).sum(),
+                torch.tensor(1.0),
+                atol=1e-3,
+            ):
+                raise ValueError("Amplitude addressing pattern weights must sum to 1.0")
+        if all([not isinstance(v, Parameter) for v in self.weights_det.values()]):
+            if not torch.isclose(
+                torch.tensor(list(self.weights_det.values())).sum(),
+                torch.tensor(1.0),
+                atol=1e-3,
+            ):
+                raise ValueError("Detuning addressing pattern weights must sum to 1.0")
+
+        # validate detuning value
+        if not isinstance(self.max_amp, Parameter):
+            if self.max_amp > DEFAULT_MAX_AMPLITUDE:
+                warn("Maximum absolute value of amplitude is exceeded")
+        if not isinstance(self.max_det, Parameter):
+            if self.max_det > DEFAULT_MAX_DETUNING:
+                warn("Maximum absolute value of detuning is exceeded")
```

Comment:
Here I suggest to compare max* max(weights) with DEFAULT_MAX, and add a test max* sum(weights) with DEFAULT_GLOBAL_MAX (value of 2000)